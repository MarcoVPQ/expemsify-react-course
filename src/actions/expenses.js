import database from '../firebase/firebase'


//ADD_EXPENSE
 export const addExpense = (expense) =>({
        type : 'ADD_EXPENSE',
        expense
    })

export const startAddExpense = (expenseData = {}) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        const { 
            description = '', 
            amount = 0, 
            createdAt = 0,
            note = ''
        } = expenseData;

        const expense = { description, amount, createdAt, note }

        return database.collection(`users/${uid}/expenses`)
        .add(expense)
        .then(docRef => {
            dispatch(addExpense({
                id: docRef.id,
                ...expense
            }))
        })
    }
}


///SET_EXPENSES
export const setExpenses = (expenses) => ({
    type: "SET_EXPENSES",
    expenses
})

export const startSetExpenses =() =>{
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.collection(`users/${uid}/expenses`)
        .get()
        .then( querySnapshot => {
            const expenses = []
            querySnapshot.forEach( doc => {
                expenses.push({
                    id: doc.id,
                    ...doc.data()
                })
            })

            dispatch(setExpenses(expenses))
        })
    }
};


//REMOVE_EXPENSE

export const removeExpense =( { id } = {} )=>({
    type: 'REMOVE_EXPENSE',
    id
})

export const startRemoveExpense =({id}) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.doc(`users/${uid}/expenses/${id}`)
              .delete()
              .then(() => {
                dispatch(removeExpense({id}))
              })
        
    }
}
//EDIT_EXPENSE

export const editExpense = (id, updates = {}) =>({
    type: "EDIT_EXPENSE",
    id,
    updates
})


export const startEditExpense = (id, updates = {}) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.doc(`users/${uid}/expenses/${id}`)
                        .update(updates)
                        .then(() => {
                            dispatch(editExpense(id, updates))
                        })
    }
}