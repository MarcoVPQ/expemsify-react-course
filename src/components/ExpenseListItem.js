import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment'



const ExpenseListItem = ({description, amount ,createdAt, id}) => (
        <Link className="list-item" to={`/edit/${id}`}>
            <div>
                <h3 className="list-item__title">{description}</h3>
                <span className="list-item__sub-title">{moment(createdAt).format('MMMM Do, YYYY')}</span>
            </div>
            <h3 className="list-item__data"> {amount}</h3>
        </Link>
);


export default ExpenseListItem;

/* install numeral@2.0.6
and using it on the amount
numeral(amount / 100).format('$0,0.00')
commit
and heroku upload*/ 