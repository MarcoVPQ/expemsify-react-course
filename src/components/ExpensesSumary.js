import React from 'react'
import { Link } from 'react-router-dom'
import { connect } from 'react-redux'

import selectExpenses from '../selectors/expenses';
import totalExpenses from '../selectors/totalExpenses'


 export class ExpensesSummary extends React.Component{
     constructor(props){
         super(props)
     }
   
    render(){
        return (
            <div className="page-header">
                <div className="content-container">
                {
                    this.props.expenses.length === 1 
                    ? ( <h2 className="page-header__title">Viewing <span>1</span> expense for a total of <span>{this.props.totalExpenses}</span></h2>)
                    : ( <h2 className="page-header__title">Viewing <span>{this.props.expenses.length}</span> expenses for a total of <span>{this.props.totalExpenses}</span></h2>)
                }
                    <div className="page-header__actions">
                        <Link className="button" to="/create">Add Expense</Link>
                    </div>
                </div>
            </div>
        )
    }
 }



const mapStateToProps = (state) => {
    return {
        expenses: selectExpenses(state.expenses, state.filters),
        totalExpenses: totalExpenses(selectExpenses(state.expenses, state.filters))
    }
}
export default connect(mapStateToProps)(ExpensesSummary)