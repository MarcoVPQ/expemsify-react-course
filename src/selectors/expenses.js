import moment from 'moment';

export default(expenses, { text, sortBy, startDate, endDate }) =>{
    return expenses.filter((expense)=>{
        const creatredAtMoment = moment(expense.createdAt);
        const startDateMatch = startDate ?  startDate.isSameOrBefore(creatredAtMoment, 'day'): true;
        const endDateMatch = endDate ? endDate.isSameOrAfter(creatredAtMoment, 'day') : true;
        const textMatch = expense.description.toLowerCase().includes(text.toLowerCase());

        return startDateMatch && endDateMatch && textMatch;
    }).sort((a,b)=>{
        if(sortBy === 'date'){
            return a.createdAt < b.createdAt ? 1 : -1;
        }else if(sortBy === 'amount'){
            return a.amount < b.amount ? 1 : -1;
        }
    })
};


