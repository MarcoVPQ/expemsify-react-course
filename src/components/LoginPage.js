import React from 'react'
import { connect } from 'react-redux'
import { startLogin } from './../actions/aut';

export const LoginPage = ({startLogin}) => (
    <div className="box-layout">
        <div className="box-layout__box">
            <h1 className="layout__title">Expensify</h1>
            <p>It's time to get your expenses in order!</p>
            <button className="button" onClick={startLogin}>Login with Google</button>
        </div>
    </div>
)

const mapDispatchToProps = (dispatch) => ({
    startLogin: () => dispatch(startLogin())
})

export default connect(undefined, mapDispatchToProps)(LoginPage)