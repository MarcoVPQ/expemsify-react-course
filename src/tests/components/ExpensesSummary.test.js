import React from 'react'
import { shallow } from 'enzyme';

import { ExpensesSummary } from '../../components/ExpensesSumary'
import expenses from '../fixtures/expenses'


test('should render Component with 1 expenses', () => {
    const wrapper = shallow(<ExpensesSummary expenses={[expenses[0]]} totalExpenses={expenses[0].amount}/>)
    expect(wrapper).toMatchSnapshot()
})

test('should render Component with 2 expenses', () => {
    const wrapper = shallow(<ExpensesSummary expenses={[expenses[0], expenses[1]]} totalExpenses={expenses[0].amount + expenses[1].amount}/>)
    expect(wrapper).toMatchSnapshot()
})