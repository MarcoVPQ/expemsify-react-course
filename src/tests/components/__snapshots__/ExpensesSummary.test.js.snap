// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Component with 1 expenses 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h2
      className="page-header__title"
    >
      Viewing 
      <span>
        1
      </span>
       expense for a total of 
      <span>
        195
      </span>
    </h2>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        replace={false}
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;

exports[`should render Component with 2 expenses 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h2
      className="page-header__title"
    >
      Viewing 
      <span>
        2
      </span>
       expenses for a total of 
      <span>
        1095195
      </span>
    </h2>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        replace={false}
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;
